game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java

```diff
@@ -131,17 +132,19 @@
   DICE_SERVER_FOR_PBEM_GAMES;
 
   private static final AtomicReference<Preferences> preferencesRef = new AtomicReference<>();
-
+  private final List<Consumer<String>> onSaveAction = new ArrayList<>();
```

Comment:
Would recommend using a `CopyOnWriteArrayList` here.  I don't think you can guarantee that `addSaveListener()` and `save()` will be called on the same thread, so this collection needs to be thread-safe.  In addition, if someone ever calls `addSaveListener()` concurrently while this collection is being iterated over in `save()`, it might trigger a `ConcurrentModificationException`.

---

Nit: Might want to pluralize the field name, too, since it is now a collection.

Reply:
>  I don't think you can guarantee that addSaveListener() and save() will be called on the same thread

No, but at same time `addSaveListener()` should always be some sort of initialization call. Notice that I did not add the 'removeListener' method as suggested, the saveAction should be immutable.

I noticed also that we call 'save' on app launch, this seems wrong, I opened a bug ticket: https://github.com/triplea-game/triplea/issues/3223

------

So, `CopyOnWriteArrayList` I don't think is quite the right answer as it implies there is a race condition. The fix for that is to not have the race. Save is expected to be driven via user action, so we should in effect be able to control the threading because of that contract.

Reply:
> No, but at same time addSaveListener() should always be some sort of initialization call. 

Then maybe we should enforce no `addSaveListener()` calls after the first `save()` call?  Otherwise, that seems like a buried assumption someone may not realize six months down the road.

> So, `CopyOnWriteArrayList` I don't think is quite the right answer as it implies there is a race condition.

The race isn't the only concern.  There's also safely publishing values between threads.  Without some sort of synchronization, there's no guarantee when the publication will happen other than "eventually."

Reply:
> Then maybe we should enforce no addSaveListener() calls after the first save() call?   Otherwise, that seems like a buried assumption someone may not realize six months down the road.

Interesting considerations. I don't think I want to lock that assumption in just yet though.